### PR TITLE
BlockPtr - replacing logical pointer with physical one

### PR DESCRIPTION
**Pull request by Horusiath**
This PR evaluates an idea of changing logical block pointers (`BlockPtr(ID, u32)`) with actual physical ones (`BlockPtr(NonNull<Block>)`). Atm. the state of the branch was just to make `yrs` into compilable state, which passes the tests and is able to perform benchmarks.

Bellow I'm pasting initial criterion results - the regressions results are vs. current `main` branch:

```
[B1.1] Append N characters/6000
                        time:   [14.487 ms 17.476 ms 19.702 ms]
                        change: [+39.451% +53.537% +61.458%] (p = 0.00 < 0.05)
                        Performance has regressed.

[B1.2] Insert string of length N/1
                        time:   [7.6715 us 7.7034 us 7.7353 us]
                        change: [+7.8507% +20.987% +36.329%] (p = 0.01 < 0.05)
                        Performance has regressed.

[B1.3] Prepend N characters/6000
                        time:   [3.2261 ms 3.2369 ms 3.2595 ms]
                        change: [-43.289% -24.700% -7.3073%] (p = 0.09 > 0.05)
                        No change in performance detected.

[B1.4] Insert N characters at random positions/6000
                        time:   [65.800 ms 65.958 ms 66.073 ms]
                        change: [-80.394% -80.202% -80.010%] (p = 0.00 < 0.05)
                        Performance has improved.

[B1.5] Insert N words at random positions/6000
                        time:   [20.851 ms 20.908 ms 20.959 ms]
                        change: [-91.284% -91.228% -91.174%] (p = 0.00 < 0.05)
                        Performance has improved.

[B1.7] Insert/Delete strings at random positions/6000
                        time:   [56.897 ms 57.009 ms 57.078 ms]
                        change: [-91.693% -91.640% -91.583%] (p = 0.00 < 0.05)
                        Performance has improved.

[B1.8] Append N numbers/6000
                        time:   [4.4764 ms 4.5460 ms 4.5983 ms]
                        change: [-15.356% -13.790% -11.651%] (p = 0.00 < 0.05)
                        Performance has improved.

[B1.9] Insert Array of N numbers/1
                        time:   [100.48 us 107.51 us 114.60 us]
                        change: [-5.7381% +6.0169% +19.915%] (p = 0.38 > 0.05)
                        No change in performance detected.

[B1.10] Prepend N numbers/6000
                        time:   [5.2612 ms 5.4252 ms 5.6646 ms]
                        change: [-15.928% -8.7977% -3.3897%] (p = 0.02 < 0.05)
                        Performance has improved.

[B1.11] Insert N numbers at random positions/6000
                        time:   [64.119 ms 64.273 ms 64.416 ms]
                        change: [-81.796% -81.688% -81.590%] (p = 0.00 < 0.05)
                        Performance has improved.

[B2.1] Concurrently insert string of length N at index 0/6000
                        time:   [34.525 us 38.571 us 40.457 us]
                        change: [+1.9310% +14.138% +28.946%] (p = 0.04 < 0.05)
                        Performance has regressed.

[B2.2] Concurrently insert N characters at random positions/6000
                        time:   [201.88 ms 202.43 ms 203.00 ms]
                        change: [-73.216% -73.044% -72.882%] (p = 0.00 < 0.05)
                        Performance has improved.

[B2.3] Concurrently insert N words at random positions/6000
                        time:   [372.67 ms 374.47 ms 376.57 ms]
                        change: [-88.446% -88.329% -88.199%] (p = 0.00 < 0.05)
                        Performance has improved.

[B2.4] Concurrently insert & delete/6000
                        time:   [201.53 ms 202.16 ms 202.71 ms]
                        change: [-87.704% -87.587% -87.476%] (p = 0.00 < 0.05)
                        Performance has improved.

[B3.1] 20√N clients concurrently set number in Map/1540
                        time:   [83.114 ms 83.161 ms 83.243 ms]
                        change: [-3.6588% -2.3349% -0.7532%] (p = 0.01 < 0.05)
                        Change within noise threshold.

Benchmarking [B3.2] 20√N clients concurrently set Object in Map/1540: Warming up for 3.0000 s
Warning: Unable to complete 10 samples in 5.0s. You may wish to increase target time to 5.3s or enable flat sampling.
[B3.2] 20√N clients concurrently set Object in Map/1540
                        time:   [96.249 ms 96.355 ms 96.530 ms]
                        change: [-6.6801% -6.4065% -6.1346%] (p = 0.00 < 0.05)
                        Performance has improved.

[B3.3] 20√N clients concurrently set String in Map/1540
                        time:   [88.898 ms 89.334 ms 89.946 ms]
                        change: [-7.3879% -6.8321% -6.2716%] (p = 0.00 < 0.05)
                        Performance has improved.

Benchmarking [B3.4] 20√N clients concurrently insert text in Array/1540: Warming up for 3.0000 s
Warning: Unable to complete 10 samples in 5.0s. You may wish to increase target time to 5.0s or enable flat sampling.
[B3.4] 20√N clients concurrently insert text in Array/1540
                        time:   [90.928 ms 91.036 ms 91.142 ms]
                        change: [+1.7003% +1.9072% +2.1317%] (p = 0.00 < 0.05)
                        Performance has regressed.

[B4.2] Apply real-world document snapshot of size/400972
                        time:   [2.4874 ms 2.4931 ms 2.4983 ms]
                        change: [+9.3238% +11.388% +13.198%] (p = 0.00 < 0.05)
                        Performance has regressed.

Benchmarking [B4.1] Apply real-world editing dataset/259778: Warming up for 3.0000 s
Warning: Unable to complete 10 samples in 5.0s. You may wish to increase target time to 54.4s.
[B4.1] Apply real-world editing dataset/259778
                        time:   [5.5071 s 5.5447 s 5.5782 s]
                        change: [-91.613% -91.521% -91.428%] (p = 0.00 < 0.05)
                        Performance has improved.
```

**tl;dr** while some tests are slower, many more of them are much faster (80-90% means 5-10x performance improvement, visible especially in *B4.1* benchmark (real-live editing trance), which takes ~60s on current main and only 5.6s on this branch). Applying document snapshot is slower, but still it takes literally milliseconds to apply.

### CODE REVIEW

**Comment by chanced** [comment]
> while some tests are slower, many more of them are much faster (80-90% means 5-10x performance improvement, visible especially in B4.1 benchmark (real-live editing trance), which takes ~60s on current main and only 5.6s on this branch).

I'm not sure I'm reading the results right, but it seems that it bailed before it finished and 54.4s was the expected runtime?

> Apply real-world editing dataset/259778: Warming up for 3.0000 s 
Warning: Unable to complete 10 samples in 5.0s. You may wish to increase target time to 54.4s.

File: tests-ffi/include/libyrs.h [hunk 1]
```diff
@@ -258,7 +258,7 @@ typedef Branch Branch;
 typedef union YOutputContent {
   char flag;
   float num;
-  long integer;
+  long long integer;
```
Comment (Horusiath) [inline tests-ffi/include/libyrs.h:261]: Since `c_long` [is variable length type](https://doc.rust-lang.org/std/os/raw/type.c_long.html) (64-bit on UNIX, 32-bit on Windows), I've changed it to `c_long_long` which makes pinky promise to always be 64-bits regardless of OS.

File: yrs/benches/benches.rs [hunk 2]
```diff
@@ -357,7 +357,7 @@ where
             let (doc, _) = iter.next().unwrap();
             let mut txn = doc.transact();
             while let Some((_, update)) = iter.next() {
-                doc.apply_update_v1(&mut txn, update.as_slice());
```
Comment (Horusiath) [inline yrs/benches/benches.rs:360]: I removed `Doc::apply_update_v1` as it's redundant - it requires a transaction scope to work anyway, and Doc itself not really necessary for this method to work. It also leaves a room for ambiguity eg. `doc1.apply_update_v1(txn2, update)` where `txn2` was potentially created from different document . Making new transaction is trivial, so I think it's a better approach.

File: yrs/src/block.rs [hunk 3]
```diff
@@ -85,55 +86,333 @@ impl ID {
...
+            Some(BranchPtr::from(branch))
+        } else {
+            None
         }
-        ptr
     }
+}
+
+impl Deref for BlockPtr {
+    type Target = Block;
```
Comment (Horusiath) [inline yrs/src/block.rs:394]: Because our smart pointer `BlockPtr` implements `Deref` over `Block` we can access its methods and fields just like we would talk to an underlying block.

File: yrs/src/block_store.rs [hunk 4]
```diff
@@ -219,24 +218,14 @@ impl ClientBlockList {
...
 
-    pub fn get(&self, index: usize) -> &Block {
-        let ptr = self.list[index].get();
-        unsafe { &*ptr }
+    pub fn try_get(&self, index: usize) -> Option<BlockPtr> {
+        let block_ref = self.list.get(index)?;
+        Some(BlockPtr::from(block_ref))
     }
 
-    pub fn get_mut(&self, index: usize) -> &mut Block {
```
Comment (Horusiath) [inline yrs/src/block_store.rs:227]: Since we're operating on `BlockPtr` which is a value type, we don't need all of the variants of `get`/`get_mut` (we can mark `BlockPtr` as `mut` at whim).

File: yrs/src/store.rs [hunk 5]
```diff
@@ -213,32 +176,27 @@ impl Store {
...
-            if let Some(item) = self.blocks.get_item_mut(&right) {
-                item.left = Some(compaction.replacement);
-            }
+        if let Some(Block::Item(right)) = compaction.new_right.as_deref_mut() {
+            right.left = Some(compaction.replacement);
         }
     }
 
-    pub fn get_type_from_path(&self, path: &Path) -> Option<BranchRef> {
+    pub fn get_type_from_path(&self, path: &Path) -> Option<BranchPtr> {
```
Comment (Horusiath) [inline yrs/src/store.rs:184]: Renamed `BranchRef` &rarr; `BranchPtr` as it's a smart pointer and a value type - obeys the same rules as `BlockPtr`.

File: yrs/src/types/text.rs [hunk 6]
```diff
@@ -128,8 +122,16 @@ impl Text {
     ///
     /// This method will panic if provided `index` is greater than the length of a current text.
     pub fn insert(&self, txn: &mut Transaction, index: u32, chunk: &str) {
-        if let Some(pos) = self.find_position(txn, index) {
+        if let Some(mut pos) = self.find_position(txn, index) {
             let value = crate::block::PrelimText(chunk.into());
+            while let Some(right) = pos.right.as_ref() {
+                if right.is_deleted() {
```
Comment (Horusiath) [inline yrs/src/types/text.rs:128]: Fix: after finding a left origin, we keep skipping over deleted blocks before inserting a new block.

File: yrs/src/id_set.rs [hunk 7]
```diff
@@ -99,14 +99,26 @@ impl IdRange {
                     if r.start > range.end {
                         *self = IdRange::Fragmented(vec![range, r.clone()])
                     } else {
-                        r.end = range.end; // two ranges overlap, we can eagerly merge them
+                        // two ranges overlap - merge them
+                        r.end = range.end.max(r.end);
+                        r.start = range.start.min(r.start);
```
Comment (Horusiath) [inline yrs/src/id_set.rs:104]: Fix for merging to delete set ranges.

File: yrs/src/transaction.rs [hunk 8]
```diff
@@ -602,21 +590,20 @@ impl Transaction {
...
                     if let Some(pivot) = list.find_pivot(r.end) {
-                        let block = &list[pivot];
-                        if block.id().clock + block.len() > r.end {
-                            blocks
-                                .split_block(&BlockPtr::new(ID::new(*client, r.end), pivot as u32));
+                        let block = list.get(pivot);
+                        let block_id = block.id();
+                        let block_len = block.len();
+                        if block_id.clock + block_len > r.end {
+                            blocks.split_block(block, block_id.clock + block_len - r.end);
```
Comment (dmonad) [inline yrs/src/transaction.rs:606]: Make sure that this method (and possibly others) use `get_item_clean_start` instead because it adds it to `merge_items` in Transaction.
